### ISSUE TITLE
[Ruby Lexer bug] Hash symbol key is sometimes regarded as error

### ISSUE THREAD

**Issue by UlyssesZh** [rouge-ruby/rouge#2009]
**Name of the lexer**
Ruby

**Code sample**
```ruby
{🎃:1}
```

**Additional context**
```ruby
require 'rouge'
Rouge::Formatters::HTML.new.format Rouge::Lexers::Ruby.lex '{🎃:1}'
```
Output:
```html
<span class=\"p\">{</span><span class=\"err\">🎃</span><span class=\"p\">:</span><span class=\"mi\">1</span><span class=\"p\">}</span>
```
The class of the `🎃` character is `err`, which is not correct (should be `n`).

**Comment by jneen** [comment]
Currently symbol names in this syntax have to start with a-z: https://github.com/rouge-ruby/rouge/blob/master/lib/rouge/lexers/ruby.rb#L86

Happy to switch this to a unicode property (`\p{...}`) if you can find documentation of what it's supposed to be.

**Comment by jneen** [comment]
The emoji does appear to work in the Ruby console though.

**Comment by jneen** [comment]
Also while I'm here this line should probably be non-greedy, even if it's protected by never matching unescaped `'`: https://github.com/rouge-ruby/rouge/blob/1687d63cede01e9e1c108425e9987060ad85c79d/lib/rouge/lexers/ruby.rb#L87

**Comment by UlyssesZh** [comment]
> Currently symbol names in this syntax have to start with a-z:
> 
> https://github.com/rouge-ruby/rouge/blob/1687d63cede01e9e1c108425e9987060ad85c79d/lib/rouge/lexers/ruby.rb#L86
> 
> Happy to switch this to a unicode property (`\p{...}`) if you can find documentation of what it's supposed to be.

It seems upper-case letters also work. My guess is that it is the same as the rules for a variable or a constant.

Does `\w` match non-ASCII characters? If no, tthe `\w*?` to match the later characters is probably not right either.

**Comment by tancnle** [comment]
Would this MR help to address this issue https://github.com/rouge-ruby/rouge/pull/1894?

**Comment by UlyssesZh** [comment]
> Would this MR help to address this issue #1894?

No. It fixes things like `啊=1` and `{啊:1}`, but not `🎃=1` or `{🎃:1}`. Seems like the new regexp rules still do not cover emojis.

**Comment by jneen** [comment]
That is curious, the regexp really should be case-insensitive as well.

**Comment by UlyssesZh** [comment]
> That is curious, the regexp really should be case-insensitive as well.

```ruby
/\p{Word}/i =~ '啊' # => 0
/\p{Word}/i =~ '🎃' # => nil
```

**Comment by tancnle** [comment]
Matching emojis seems to be tricky as they can be captured in a regex range. I believe this gem https://github.com/ticky/ruby-emoji-regex might have the regex to solve it 🤔  

```pry
[1] pry(main)> require 'emoji_regex'
[2] pry(main)> EmojiRegex::Regex.match('🎃')
=> #<MatchData "🎃">
```

**Comment by UlyssesZh** [comment]
> Matching emojis seems to be tricky as they can't be captured in a regex range. I believe this gem https://github.com/ticky/ruby-emoji-regex might have the regex to solve it 🤔
> 
> ```
> [1] pry(main)> require 'emoji_regex'
> [2] pry(main)> EmojiRegex::Regex.match('🎃')
> => #<MatchData "🎃">
> ```

That does not solve the issue either because there are non-word characters other than emojis that can start a name, such as the Chinese period:
```ruby
/\p{Word}/i =~ '。' # => nil
。=1 # no SyntaxError
```
I think any non-ASCII characters can do it? So what we really need to use is `[a-z_|^\x00-\x7F]` (for variables) and `[a-zA-Z_|^\x00-\x7F]` (for Hash symbol keys).

**Comment by jneen** [comment]
I'm a little nervous about significantly expanding the name rules without some assurance from Ruby folks about what it is exactly they intend to support. Does ruby have a spec for this?

**Comment by UlyssesZh** [comment]
> I'm a little nervous about significantly expanding the name rules without some assurance from Ruby folks about what it is exactly they intend to support. Does ruby have a spec for this?

There should have a spec in [ISO/IEC 30170:2012](https://www.iso.org/standard/59579.html), but I didn't buy it. [The documentation on ruby-doc.org](https://ruby-doc.org/3.2.2/syntax/assignment_rdoc.html#label-Local+Variable+Names) says any character with the eighth bit set can start a variable, though, so `[^\x00-\x7F]` should be it.

**Comment by jneen** [comment]
<img width="1029" alt="image" src="https://github.com/rouge-ruby/rouge/assets/225017/09e3ae43-ae74-46c0-a15e-2a904c320a9b">
Looks right to me. I didn't want to believe it but it looks like they *actually* implemented it this way:

``` ruby
; ruby -e 'p 　:1'
{:　=>1}
```

This is a double-width space, also known as `\xe3\x80\x80`, and it can be used as an identifier 🤦 

As long as there aren't any issues with that regexp leaving glyphs stranded, we should be able to support this somewhat. Basic testing with the Japanese block seems to indicate that `[^\x00-\x7F]` will work.

**Comment by UlyssesZh** [comment]
![image](https://github.com/rouge-ruby/rouge/assets/26196187/22f2436b-eb9d-4eb6-940e-530dbf94c1ee)

It seems that ISO/IEC 30170:2012 does not talk about non-ASCII characters, so **strictly speaking** using non-ASCII characters as identifiers is undefined behavior. It just happens to be implemented and documented in CRuby and maybe other Ruby implementations.

I do not know what is the philosophy of Rouge, but I am supporting adopt `[^\x00-\x7F]`. I have read some of the ISO documentation, and I must say it just gave *way* too much freedom to implementations of Ruby interpreters. CRuby should be used as the *de facto* standard instead (and it actually is, when people try to invent new implementations), and any Ruby highlighting program should also use CRuby as the standard.

**Comment by jneen** [comment]
Looks like at least one other implementation agrees:

https://github.com/lib-ruby-parser/lib-ruby-parser/blob/134edd54bac26dee604e27ea6d5537c20b265646/src/source/buffer.rs#L316
